### PR TITLE
feat(form-field): implement component

### PR DESCRIPTION
**Pull request by Leotheluck**
## A Form Field component is used to wrap several form components with logic, visual hints and additional styling.

### CODE REVIEW

File: packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss [hunk 1]
```diff
@@ -0,0 +1,24 @@
+:host(.ods-form-field) {
+  display: inline-flex;
+  position: relative;
```
Comment (dpellier) [inline packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss:3]: should be useless

File: packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+.ods-form-field {
+  &__top-bar {
+    display: flex;
+    flex-direction: row;
+    align-items: flex-end;
+    justify-content: space-between;
+    margin-bottom: 4px;
+    width: 100%;
```
Comment (dpellier) [inline packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss:14]: should be useless

File: packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss [hunk 3]
```diff
@@ -0,0 +1,24 @@
+:host(.ods-form-field) {
+  display: inline-flex;
+  position: relative;
+  flex-direction: column;
+}
+
+.ods-form-field {
+  &__top-bar {
+    display: flex;
+    flex-direction: row;
```
Comment (dpellier) [inline packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss:10]: add a `column-gap` to ensure spacing if label is long enough to reach the second element

File: packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    align-items: flex-end;
+    justify-content: space-between;
+    margin-bottom: 4px;
```
Comment (dpellier) [inline packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss:13]: remove, use a `row-gap` on parent instead

File: packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    width: 100%;
+  }
+
+  &__bottom-bar {
+    margin-top: 4px;
```
Comment (dpellier) [inline packages/ods/src/components/form-field/src/components/ods-form-field/ods-form-field.scss:18]: remove, use a `row-gap` on parent instead

File: packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    await setup('<ods-form-field error=""><ods-input></ods-input></ods-form-field>');
+    const errorMessage = await page.find('ods-form-field >>> .ods-form-field__bottom-bar__error-message');
+    expect(errorMessage).toBeNull();
+  });
+
+  it('should correctly render content in the label slot', async() => {
+    await setup('<ods-form-field><ods-input></ods-input><span slot="label">Label</span></ods-form-field>');
+    const labelSlot = await el.find('span[slot="label"]');
+    expect(labelSlot).not.toBeNull();
+    expect(labelSlot.textContent).toEqual('Label');
```
Comment (dpellier) [inline packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts:43]: `toBe`

File: packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  });
+
+  it('should correctly render content in the visual-hint slot', async() => {
+    await setup('<ods-form-field><ods-input></ods-input><span slot="visual-hint">Visual Hint</span></ods-form-field>');
+    const visualHintSlot = await el.find('span[slot="visual-hint"]');
+    expect(visualHintSlot).not.toBeNull();
+    expect(visualHintSlot.textContent).toEqual('Visual Hint');
```
Comment (dpellier) [inline packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts:50]: `toBe`

File: packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    await setup('<ods-form-field><ods-textarea></ods-textarea></ods-form-field>');
+    const textarea = await el.find('ods-textarea');
+    expect(textarea).not.toBeNull();
+  });
+
+  it('should correctly render helper text in the helper slot if no error', async() => {
+    await setup('<ods-form-field><ods-input></ods-input><span slot="helper">Helper</span></ods-form-field>');
+    const helperSlot = await el.find('span[slot="helper"]');
+    expect(helperSlot).not.toBeNull();
+    expect(helperSlot.textContent).toEqual('Helper');
```
Comment (dpellier) [inline packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts:69]: `toBe`

File: packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  });
+
+  it('should be displaying the error message if error, even if there is a helper', async() => {
+    await setup('<ods-form-field error="Wrong format."><ods-input></ods-input><span slot="helper">Helper Text</span></ods-form-field>');
+    const errorMessage = await page.find('ods-form-field >>> .ods-form-field__bottom-bar__error-message');
+    expect(errorMessage).not.toBeNull();
+    expect(errorMessage?.textContent).toEqual('Wrong format.');
```
Comment (dpellier) [inline packages/ods/src/components/form-field/tests/rendering/ods-form-field.e2e.ts:76]: `toBe`

File: packages/storybook/stories/components/form-field/documentation.mdx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+<Canvas of={ FormFieldStories.Default } sourceState='none' />
+
+<DocNavigator of={ FormFieldStories } />
+
+<Markdown>{ SpecificationsFormField }</Markdown>
+
+# Style customization
+
+You can add your own style on the form-field elements using `::part` or by directly updating the slot elements.
```
Comment (dpellier) [inline packages/storybook/stories/components/form-field/documentation.mdx:23]: feels weired to mention parts here as they can use whatever they want in the form-field, not always ods with parts

File: packages/storybook/stories/components/form-field/form-field.stories.ts [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  args: {
+    error: '',
+    label: `<ods-text slot="label" preset='label'>Description</ods-text>`,
+    visualHint: `<ods-text slot="visual-hint" preset='caption'>02/11/1999</ods-text>`,
+    mainSlot: `<ods-input type="text" value="Hello, ODS!" clearable></ods-input>`,
+    helper: `<ods-text slot="helper" preset='span'>A little helper text</ods-text>`,
+  },
+};
+
+export const Default: StoryObj = {
```
Comment (dpellier) [inline packages/storybook/stories/components/form-field/form-field.stories.ts:76]: Default would be the bare minimum required, in this case only input
Maybe use one of the complete one for the top doc page overview to have a better exemple and update this one?

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by feoche — CHANGES_REQUESTED** [review]
### Documentation
- [x] Style customization : prefer something less flashy (i.e. #b63f81 as text color)
- [x] Style customization : remove visual hint slot
- [x] Examples : put slotted examples above "Error" one
- [x] Examples : Tooltip → Use `help-circle` icon instead of "(???)"

### Demo
- [x] Change `mainSlot` control into `content`, no description (only `string` appears)
- [x] Default content
  - [x] Label should be 14px semibold / #4d5592
  - [x] Visual hint should be 12px regular / #4d5592
  - [x] Helper message should be 12px regular / #4d5592
  - [x] Error message should be 12px semi-bold #bf0020

### Style
- [x] Visual hint slot should be in bottom right part (see example below)
image

**Review by dpellier — CHANGES_REQUESTED** [review]
Style customization does not work
Capture d’écran 2024-05-15 à 17 30 34

**Review by dpellier — CHANGES_REQUESTED** [review]
Change textarea example to a number of character typed value instead, like `11/200`
Capture d’écran 2024-05-15 à 17 31 38

**Review by feoche — CHANGES_REQUESTED** [review]
### General
- [ ] Set input and textarea with "Placeholder" `placeholder` only, no value
- [x] Set visual hint text to `0/100`

### Documentation
- [x] Put Tooltip icon inside "label" slot, at the end

### Demo
- [x] Change `caption` visual hint into `span` preset

**Comment by feoche** [comment]
[ ] Fix conflicts

**Review by dpellier — APPROVED** [review]

**Review by feoche — APPROVED** [review]

**Review by astagnol — APPROVED** [review]

**Pull request merged**